core/chains/chain_kv.go

```diff
@@ -0,0 +1,85 @@
+package chains
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+)
+
+type ChainsKV[T ChainService] struct {
+	mu     sync.Mutex
+	chains map[string]T
+}
+
+var ErrNoSuchChainID = errors.New("chain id does not exist")
+
+func NewChainsKV[T ChainService]() *ChainsKV[T] {
+	return &ChainsKV[T]{
+		chains: map[string]T{},
+	}
+}
+func (c *ChainsKV[T]) Len() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return len(c.chains)
+}
+func (c *ChainsKV[T]) lazyInit() {
+	if c.chains == nil {
+		c.chains = make(map[string]T)
+	}
+}
```

Comment:
nit/ it seems like this could be eliminated if the constructor were the only way to create a `ChainsKV`.